presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -173,22 +168,22 @@ private void onMemoryReserved()
     public boolean tryReserve(QueryId queryId, String allocationTag, long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
+        if (getFreeBytes() - bytes < 0) {
+            return false;
+        }
         synchronized (this) {
-            if (getFreeBytes() - bytes < 0) {
-                return false;
-            }
             reservedBytes += bytes;
-            if (bytes != 0) {
-                queryMemoryReservations.merge(queryId, bytes, Long::sum);
-                updateTaggedMemoryAllocations(queryId, allocationTag, bytes);
-            }
+        }
+        if (bytes != 0) {
+            queryMemoryReservations.merge(queryId, bytes, Long::sum);
+            updateTaggedMemoryAllocations(queryId, allocationTag, bytes);
         }
 
         onMemoryReserved();
         return true;
     }
 
-    public synchronized void free(QueryId queryId, String allocationTag, long bytes)
+    public void free(QueryId queryId, String allocationTag, long bytes)
```

Comment:
I don't think this is safe.

There is a race between a thread freeing memory and reducing the reservation to 0 and a thread adding memory.

Thread A could reserve memory and then have it removed by this function in thread B because B saw the the reservation hit 0, but then doesn't run while A is adding to the reservation.

The remove needs to be conditional on the queryReservationValue that is there not being changed.

Generally the approach to removing the memory reservation on free is fragile. We should remove the reservation information when the query itself is cleaned up/removed from the node.